### PR TITLE
add db_connection package

### PR DESCRIPTION
**Pull request by AtanasKisyov**

### CODE REVIEW

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    def __init__(self, hostname, database_name, pool_size: int) -> None:
+        """Initialize read-only pool"""
+
+        self.hostname = hostname
+        self.database_name = database_name
+        self._pool = deque()
+        self.pool_size = pool_size
+        self._open_connection = None
+    
+    def __enter__(self) -> psycopg2.connect:
```
Comment (ulyktey) [inline Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py:24]: Please replace __enter__ and __exit__ into my last example in chat

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/pool_base_class.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import psycopg2
+
+from python.car_wash.utils.singleton import Singleton
+
+
+class PoolBaseClass(metaclass=Singleton):
+
+    """Abstract base class for creating connection pools"""
+
+    def __enter__(self) -> psycopg2.connect:
```
Comment (ulyktey) [inline Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/pool_base_class.py:11]: Based on my last version of example you will not have enter and exit

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
+"""Read-only and read/write database connection pools definitions"""
+from collections import deque
+import logging
+import time
+
+import psycopg2
+
+from python.car_wash.db_connection.pool_base_class import PoolBaseClass
+
+
```
Comment (ulyktey) [inline Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py:10]: Create constant as i showed https://github.com/ulyktey/car_wash_db_pool/blob/main/src/python/car_wash/db/dbpool.py#L10 and use it for class parameters

**Review by ulyktey — CHANGES_REQUESTED** [review]

**Comment by ulyktey** [comment]
Please rewrite code based on my example: https://github.com/ulyktey/car_wash_db_pool/blob/main/src/python/car_wash/db/dbpool.py

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        self.hostname = hostname
+        self.database_name = database_name
+        self._pool = deque()
+        self.pool_size = pool_size
+        self._open_connection = None
+    
+    def __enter__(self) -> psycopg2.connect:
+        """Get database connection from pool"""
+
+        if not self._pool:
```
[the remark this answers is not in the document]
Comment (AtanasKisyov) [reply Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py]: Done.

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py (hunk 1 above)
Comment (AtanasKisyov) [reply Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py]: Done.

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/pool_base_class.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+"""Abstract base class to ensure that database pools implement the correct methods"""
+import psycopg2
+
+from python.car_wash.utils.singleton import Singleton
+
+
+class PoolBaseClass(metaclass=Singleton):
```
[the remark this answers is not in the document]
Comment (AtanasKisyov) [reply Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/pool_base_class.py]: Done.

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/pool_base_class.py (hunk 2 above)
Comment (AtanasKisyov) [reply Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/pool_base_class.py]: Done.

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+import psycopg2
+
+from python.car_wash.db_connection.pool_base_class import PoolBaseClass
+
+
+class ReadOnlyPool(PoolBaseClass):
+
+    """Read-only connection pool"""
+
+    def __init__(self, hostname, database_name, pool_size: int) -> None:
```
[the remark this answers is not in the document]
Comment (AtanasKisyov) [reply Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py]: Done.

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py (hunk 3 above)
Comment (AtanasKisyov) [reply Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py:12]: Done.

**Comment by AtanasKisyov** [comment]
Everything from the comments and the change request is implemented/changed. As requested, this is the output of the execution:
```
python3 main.py -ro
INFO:root:Connection successful!
INFO:root:Connection successful!
INFO:root:Connection successful!
INFO:root:Connection successful!
INFO:root:Connection successful!
INFO:root:5 connections are added to the pool!
INFO:root:Connection successful!
INFO:root:Connection successful!
INFO:root:2 connections are added to the pool!
INFO:root:A free connection is pulled from the pool!
(1, '+359-00-000-000', 'Arthur Dent', 'Earth')
(2, '+359-11-111-111', 'Slartibartfast', 'Magrathea')
(3, '+359-22-222-222', 'Ford Prefect', 'Earth')
INFO:root:Connection has been successfuly returned to the pool!
```

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            logging.error('There are no free connections at the moment. Please wait for free connection...')
+            time.sleep(1)
+
+        # Get connection from pool and log it is successful
+        self._open_connection = self._pool.popleft()
+        logging.info('A free connection is pulled from the pool!')
+
+        yield self._open_connection
+
+        self._open_connection.commit()
```
Comment (ulyktey) [inline Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py:43]: We need to commit for WRITE type only!!!

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    def connection(self) -> typing.Generator:
+        """Get database connection from pool, yield it, and return it to the pool after it's no longer used"""
+
+        # If there are no connections in the pool, wait until one is returned
+        while not self._pool:
+            logging.error('There are no free connections at the moment. Please wait for free connection...')
+            time.sleep(1)
+
+        # Get connection from pool and log it is successful
+        self._open_connection = self._pool.popleft()
```
Comment (ulyktey) [inline Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py:38]: You do not need self here. you operate by local variable

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        logging.info('A free connection is pulled from the pool!')
+
+        yield self._open_connection
+
+        self._open_connection.commit()
+        self._pool.append(self._open_connection)
+        self._open_connection = None
+        logging.info('Connection has been successfuly returned to the pool!')
+
```
Comment (ulyktey) [inline Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py:47]: Please create method get_connection:
it should contain:
```
connection = None
while not connection:
        with lock:
                if -> check if we have some connection in pool.
                yes: get connection from pool
                no: check that we can open new connection
                log.info('No free connection in pool. Wait 0.1 second.')
                time.sleep(0.1)
return connection
                 
```

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+            self._pool.append(connection)
+
+        log_message = f'{self.pool_size} connections are added to the pool!'
+        logging.info(log_message)
+
+
+class ReadOnlyPool(PoolBaseClass):
+
+    """Read-only connection pool"""
+    pass
```
Comment (ulyktey) [inline Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py:79]: You will have different `connection` method

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
+from db_connection.config import USER_CREDENTIALS
+from db_connection.connection_pools import ReadOnlyPool, ReadWritePool
+from utils.parse_arguments import add_arguments
+
+
+def main():
+    """Interact with user"""
+
```
Comment (ulyktey) [inline Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py:8]: all operations with argument should be in main method.

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    #  Get arguments from user command
+    connection_options = add_arguments()
+    arg = {'arg': key for key, value in connection_options.items() if value}
```
Comment (ulyktey) [inline Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py:11]: You do not need to do this

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    arg = arg.get('arg', 'ro')
+
+    # Create connection pools
+    read_only_pool = ReadOnlyPool(**USER_CREDENTIALS, pool_size=5)
+    read_write_pool = ReadWritePool(**USER_CREDENTIALS, pool_size=2)
+
+    # Populate the connection pools
+    read_only_pool.populate_pool()
```
Comment (ulyktey) [inline Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py:19]: We create connection when we do DB query. Please remove this method

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/utils/parse_arguments.py [hunk 14]
```diff
@@ -0,0 +1,12 @@
+import argparse
+import logging
+
+
+def add_arguments() -> dict:
+    """Initialize available arguments for user to pass to the program"""
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-rw', '--rw', action='store_true')
```
Comment (ulyktey) [inline Task3/Task3_Atanas_Kisyov/src/python/car_wash/utils/parse_arguments.py:9]: Please add the default value of pool size. Please add description

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/utils/credentials.py [hunk 15]
```diff
@@ -0,0 +1,16 @@
+def read_user_credentials(file_path: str) -> list:
```
Comment (ulyktey) [inline Task3/Task3_Atanas_Kisyov/src/python/car_wash/utils/credentials.py:1]: We do not need this but it is not bad. Please remove it.

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    with choosen_connection.connection() as conn:
+        cursor = conn.cursor()
+        cursor.execute('SELECT * FROM users')
+        table = cursor.fetchall()
+        for row in table:
+            print(row)
+
+
+if __name__ == '__main__':
+    main()
```
Comment (ulyktey) [inline Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py:38]: sys.exit(main)

**Review by ulyktey — CHANGES_REQUESTED** [review]
Please check my comments

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py (hunk 11 above)
Comment (AtanasKisyov) [reply Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py:8]: I am not sure I understand. Can you provide more information, please?

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py (hunk 12 above)
Comment (AtanasKisyov) [reply Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py:11]: How else should I get the passed argument? Or do you mean to move it in add_argument function?

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/utils/parse_arguments.py (hunk 14 above)
Comment (ulyktey) [reply Task3/Task3_Atanas_Kisyov/src/python/car_wash/utils/parse_arguments.py:9]: Let's remove argparse and do our code easier

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py (hunk 12 above)
Comment (ulyktey) [reply Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py:11]: `if arg.ro: ...`
But Let's remove argparse and do our code easier

Comment (AtanasKisyov) [reply Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py:11]: sys.argv ?

Comment (ulyktey) [reply Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py:11]: We will skip this functionality and spend time to fixe all other issues

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+    with read_write_pool.connection() as conn:
+        cursor = conn.cursor()
+        cursor.execute('INSERT...')
+
+    with read_only_pool.connection() as conn:
+        cursor = conn.cursor()
+        cursor.execute('SELECT * FROM users')
+        table = cursor.fetchall()
+        for row in table:
+            print(row)
```
Comment (ulyktey) [inline Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py:24]: logging should be here

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py (hunk 11 above)
Comment (ulyktey) [reply Task3/Task3_Atanas_Kisyov/src/python/car_wash/main.py:10]: Skip this comment

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+            logging.critical('Connection failed! \
+                See exception message and try again!')
+            raise psycopg2.OperationalError(exception)
+
+    def get_connection(self) -> psycopg2.connect:
+        """Get connection from the pool"""
+
+        # Create and return connection
+        # If the count is less than the pool max size
+        if self.pool_size > 0:
```
Comment (ulyktey) [inline Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py:56]: What about multiprocess? 
Please try to add `lock`. Check my recommendation with lock :)

**Review by ulyktey — COMMENTED** [review]
Please check my comments. Globaly ok but you forgot about multy processes

File: Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py (hunk 18 above)
Comment (ulyktey) [reply Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py:58]: You need to merge your 58 line + 63-66 + 69-71 lines :). Your code has issue with multithreading. Lock is code that allows only access for 1 thread at 1 time. It means  `lock` should be in the same loop as all other checks. Please check my recommendation about  this.

Comment (AtanasKisyov) [reply Task3/Task3_Atanas_Kisyov/src/python/car_wash/db_connection/connection_pools.py:58]: If I understand you correctly, I should wrap the whole method in lock?

**Pull request merged**